dips/dip-196.md

```diff
@@ -0,0 +1,70 @@
+---
+dip: 196
+title: Onchain Network Identity
+authors: Greg Nazario (@gregnazario)
+status: Draft
+type: Informational
+created: 07/06/2021
+last_updated: 07/06/2021
+issue: https://github.com/diem/dip/issues/196
+---
+
+import useBaseUrl from '@docusaurus/useBaseUrl';
+
+# Simple Summary
+
+This DIP describes Onchain Network, an identity for Diem nodes to easily discover and update identity of other Diem nodes on the network.  Allows for trusted connections between nodes.
+
+# Abstract
+
+Currently, network discovery is done either manually (via configuration) or by the validator set (which only applies to validators and their respective full nodes).  Adding onchain network identity allows for nodes to discover other nodes via the shared onchain resources, and allow owners of the nodes to control their identity and updates to the identity. This DIP describes a method of storing a node's `Network Identity` on-chain for use by other nodes in node discovery. A `NetworkIdentity` is a combination of a `PublicKey` and an `AccountAddress` that represents the node. An `AccountAddress` is derived from the `PublicKey` and should be unique per running node. This `NetworkIdentity` will be stored on chain per owning account e.g. a VASP.  This `NetworkIdentity` will be used by other nodes to identify other nodes to allow connections by subscribing to events for specific accounts.
+
+
+# Motivation
+
+This method will be used first for the Association to connect their full nodes to the validator full node network, with uninterrupted access.  Other nodes may choose to use this method for discovery in the future.
+
+# Specification
+
+## Onchain Network Identity Format
+
+The identities shall be a vector of bcs encoded `NetworkIdentity` objects.  However, the vector shall be a set, and any duplicates will be ignored.
+```
+{
+  identities: vector<vector<u8>>,
+}
+```
+
+The `NetworkIdentity` struct shall be a tuple of `peer_id` and a x25519 `public_key` that represents the identity for a single node.
+```
+{
+    peer_id: <account_address>,
+    public_key: <base64 encoded x25519 public key>,
+}
+```
+
+## Updating Network Identity - Owner
+
+The `NetworkIdentity` for an account shall be updated by only the owner of the account.  At this point, the consumers of this identity must update their local use of identity within a reasonable amount of time.
+
+## Consuming Network Identity
+
+Consumers of `NetworkIdentity` may keep track of a list of accounts to watch.  The consumers then can watch the event streams of each account's `NetworkIdentity` for updates and update accordingly.
+
+A consumer may choose to subscribe to all accounts, in which it will watch the `Treasury compliance` account updates and keep track of which accounts to watch accordingly.
```

Comment:
can you provide the code snippet?

Reply:
Currently reworking some of the technical details involving integration with move.  Can come up with examples and code snippets once that's done.